### ISSUE TITLE
validating JsonApiDateTime field

### ISSUE THREAD

**Issue by dreamsbond** [limoncello-php/app#7]
in Model setting:
self::FIELD_DATELINE      => JsonApiDateTimeType::NAME,

with Validator setting:
Scheme::ATTR_DATELINE => $this->isDateTime(),

in test:

        {
            "data": {
                "type": "holidays",
                "id": null,
                "attributes": {
                	"dateline": "2016-06-15T01:38:37+0000"
                }
            }
        }

and return:

Neomerx\JsonApi\Exceptions\ErrorCollection Object
(
    [items:Neomerx\JsonApi\Exceptions\ErrorCollection:private] => Array
        (
            [0] => Neomerx\JsonApi\Document\Error Object
                (
                    [idx:Neomerx\JsonApi\Document\Error:private] =>
                    [links:Neomerx\JsonApi\Document\Error:private] =>
                    [status:Neomerx\JsonApi\Document\Error:private] => 422
                    [code:Neomerx\JsonApi\Document\Error:private] =>
                    [title:Neomerx\JsonApi\Document\Error:private] => Invalid element.
                    [detail:Neomerx\JsonApi\Document\Error:private] => The `dateline` value is invalid.
                    [source:Neomerx\JsonApi\Document\Error:private] => Array
                        (
                            [pointer] => /data/attributes/dateline
                        )

                    [meta:Neomerx\JsonApi\Document\Error:private] =>
                )

        )

)
 seems like the validator does not parse the 2016-06-15T01:38:37+0000 while using $this->isDateTimeFormat

**Comment by neomerx** [comment]
@dreamsbond There are 2 issues here. One is yours and one is mine :smile: 

You have a string as an input value and validate it as a DateTime which predictably fails.
Instead, you have to have your validation rules something like
```php
       Scheme::ATTR_DATELINE   => $this->dateline(),
```
and the rule itself
```php
    protected function dateline(): RuleInterface
    {
        return $this->toDateTime($this->success(), DateBaseType::JSON_API_FORMAT);
    }
```
Instead of `$this->success()` you're likely to have some range check for your dates.

As for mine error, I haven't implemented proper parsing for input dates. A quick fix for that would be adding the following method to `\Limoncello\Flute\Types\JsonApiDateTimeType` (`vendor/limoncello-php/flute/src/Types/JsonApiDateTimeType.php`)

```php

    /**
     * @inheritdoc
     */
    public function convertToDatabaseValue($value, AbstractPlatform $platform)
    {
        assert(is_string($value));
        $dateTime = \DateTimeImmutable::createFromFormat(DateBaseType::JSON_API_FORMAT, $value);
        assert($dateTime !== false);

        return parent::convertToDatabaseValue($dateTime, $platform);
    }
```

A proper fix will be added and published later this weekend.

**Comment by neomerx** [comment]
fixed in `0.6.15`

Don't hesitate to contact me if you have any further questions.
